### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by Amir-Gaifullin**
Эту ошибку не могу исправить в edit comment, написал всё вроде правильно но никак.
image

### CODE REVIEW

**Comment by Amir-Gaifullin** [comment]
Проблема решена путём добавление функционала в Comments => edit

File: app/controllers/tasks_controller.rb [hunk 1]
```diff
@@ -48,6 +48,7 @@ def destroy
     redirect_to tasks_path
   end
 
+
```
Comment (DmitryBarskov) [inline app/controllers/tasks_controller.rb]: 👍

File: app/controllers/projects_controller.rb [hunk 2]
```diff
@@ -30,6 +30,7 @@ def create
     @project = Project.new(project_params)
 
     if @project.save
+      puts project_params
```
Comment (DmitryBarskov) [inline app/controllers/projects_controller.rb]: Remove this line

File: app/views/tasks/_form.html.erb [hunk 3]
```diff
@@ -17,5 +14,8 @@
   <%= form.label :deadline_at, 'Deadline At' %>
   <%= form.datetime_local_field :deadline_at %>
 
+  <%= form.label :status, 'Select a status task' %>
```
Comment (DmitryBarskov) [inline app/views/tasks/_form.html.erb]:
```suggestion
  <%= form.label :status, 'Select a task status' %>
```

File: app/controllers/comments_controller.rb [hunk 4]
```diff
@@ -1,28 +1,48 @@
...
   end
 
   def update
-    if comment.update(comment_params)
-      redirect_to comment, notice: 'Comment was successfully updated.'
+    if @comment.update(comment_params)
+      redirect_to task_path(@comment.task_id), notice: 'Comment was successfully updated.'
     else
-      redirect_to comment.task, alert: 'Comment was not updated'
+      redirect_to tasks_path, alert: 'Comment was not updated'
```
Comment (DmitryBarskov) [inline app/controllers/comments_controller.rb]: But how user knows what's wrong?

File: app/models/comment.rb [hunk 5]
```diff
@@ -1,6 +1,4 @@
 class Comment < ApplicationRecord
-  belongs_to :project
```
Comment (DmitryBarskov) [inline app/models/comment.rb:2]: You can still have it using `through` option:
```ruby
belongs_to :project, through: :task
```

File: app/policies/project_policy.rb [hunk 6]
```diff
@@ -14,7 +14,7 @@ def create?
   end
 
   def update?
-    user.id == project.user_id
+    project.user_id.in?(project.users.ids)
```
Comment (DmitryBarskov) [inline app/policies/project_policy.rb]: Don't use `in?` (https://rails.rubystyle.guide/#active_support_extensions)
```suggestion
    project.user_ids.include?(project.user_id)
```

File: app/views/comments/_form.html.slim [hunk 7]
```diff
@@ -0,0 +1,11 @@
+ 
+p
+  strong Add Comment
+
+= simple_form_for [@task, @comment] do |form|
+  = form.label :task_id, 'Select a task'
+  = form.select :task_id, Task.all.map{|task|[task.title, task.id]}
```
Comment (DmitryBarskov) [inline app/views/comments/_form.html.slim]: Use `pluck` here
```suggestion
  = form.select :task_id, Task.all.pluck(:title, :id)
```

File: app/views/comments/_list.html.slim [hunk 8]
```diff
@@ -0,0 +1,10 @@
+| List comments: 
+- @comments.each do |comment|
+	p
+  	= User.find(comment.user_id).full_name
```
Comment (DmitryBarskov) [inline app/views/comments/_list.html.slim]: Why not just `comment.user.full_name`?

**Review by DmitryBarskov — CHANGES_REQUESTED** [review]

File: app/views/tasks/_form.html.erb (hunk 3 above)
Comment (Amir-Gaifullin) [reply app/views/tasks/_form.html.erb]: kek

File: app/models/comment.rb (hunk 5 above)
Comment (Amir-Gaifullin) [reply app/models/comment.rb:2]: why this better than just ```belongs_to :project```

File: app/views/comments/_list.html.slim (hunk 8 above)
Comment (Amir-Gaifullin) [reply app/views/comments/_list.html.slim]: probably didn't think about it))

File: app/controllers/comments_controller.rb (hunk 4 above)
Comment (Amir-Gaifullin) [reply app/controllers/comments_controller.rb]: agree

File: app/models/comment.rb (hunk 5 above)
Comment (Amir-Gaifullin) [reply app/models/comment.rb:2]: this option don't have in my rails. He just throws me a mistake ```Unknown key: :through```

Comment (Amir-Gaifullin) [reply app/models/comment.rb:2]: Why do we need to say the model of comments belongs to project, when they belong only to task, I don't understand?

Comment (DmitryBarskov) [reply app/models/comment.rb:2]: ok

File: app/models/comment.rb [hunk 9]
```diff
@@ -1,8 +1,8 @@
...
-
   validates :content, presence: true
 
   belongs_to :user
   belongs_to :task
+  belongs_to :project, :through => :task
```
Comment (DmitryBarskov) [inline app/models/comment.rb]: Then delete it

File: db/migrate/20211028124959_add_user_to_task.rb [hunk 10]
```diff
@@ -0,0 +1,5 @@
+class AddUserToTask < ActiveRecord::Migration[6.1]
+  def change
+    add_column :tasks, :user_id, :integer, null: true
```
Comment (DmitryBarskov) [inline db/migrate/20211028124959_add_user_to_task.rb]: Use `reference` instead

File: app/views/comments/_form.html.slim [hunk 11]
```diff
@@ -0,0 +1,11 @@
+ 
+p
+  strong Add Comment
+
+= simple_form_for [@task, @comment] do |form|
+  = form.label :task_id, 'Select a task'
+  = form.select :task_id, Task.all.pluck(:title, :id)
```
Comment (DmitryBarskov) [inline app/views/comments/_form.html.slim]: User don't need to select a task, since you already have `@task`

**Review by DmitryBarskov — CHANGES_REQUESTED** [review]

File: app/models/comment.rb (hunk 9 above)
Comment (Amir-Gaifullin) [reply app/models/comment.rb]: sec

File: db/migrate/20211028124959_add_user_to_task.rb (hunk 10 above)
Comment (Amir-Gaifullin) [reply db/migrate/20211028124959_add_user_to_task.rb]: это код Миши

Comment (DmitryBarskov) [reply db/migrate/20211028124959_add_user_to_task.rb]: Then request review from him

Comment (KilinMikhail) [reply db/migrate/20211028124959_add_user_to_task.rb]: @DmitryBarskov I saw previuos code with adding user to project. It seems like you decided to not use reference and I decided to keep it simple

File: app/controllers/comments_controller.rb [hunk 12]
```diff
@@ -1,28 +1,54 @@
 class CommentsController < ApplicationController
   before_action :authenticate_current_user!
   before_action :set_comment
+  before_action -> { authorize @comment }, only: %i[edit update destroy]
```
Comment (KilinMikhail) [inline app/controllers/comments_controller.rb:4]:
```suggestion
  before_action -> { authorize @comment }, only: %i[edit update destroy new]
```
Why do you do it in new action?

**Review by KilinMikhail — APPROVED** [review]

File: app/controllers/comments_controller.rb (hunk 12 above)
Comment (Amir-Gaifullin) [reply app/controllers/comments_controller.rb:4]: create comment?

Comment (KilinMikhail) [reply app/controllers/comments_controller.rb:4]: new

Comment (KilinMikhail) [reply app/controllers/comments_controller.rb:4]: it is not principal, it strange)

**Pull request closed without merging**